sonar-core/src/main/java/org/sonar/core/issue/tracking/BlockRecognizer.java

```diff
@@ -0,0 +1,177 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.core.issue.tracking;
+
+import com.google.common.collect.LinkedHashMultimap;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Multimap;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+class BlockRecognizer<RAW extends Trackable, BASE extends Trackable> {
+
+  /**
+   * If base source code is available, then detect code moves through block hashes.
+   * Only the issues associated to a line can be matched here.
+   */
+  void match(Input<RAW> rawInput, Input<BASE> baseInput, Tracking<RAW, BASE> tracking) {
+    BlockHashSequence rawHashSequence = rawInput.getBlockHashSequence();
+    BlockHashSequence baseHashSequence = baseInput.getBlockHashSequence();
+
+    Multimap<Integer, RAW> rawsByLine = groupByLine(tracking.getUnmatchedRaws(), rawHashSequence);
+    Multimap<Integer, BASE> basesByLine = groupByLine(tracking.getUnmatchedBases(), baseHashSequence);
+    Map<Integer, HashOccurrence> occurrencesByHash = new HashMap<>();
+
+    for (Integer line : basesByLine.keySet()) {
+      int hash = baseHashSequence.getBlockHashForLine(line);
+      HashOccurrence hashOccurrence = occurrencesByHash.get(hash);
+      if (hashOccurrence == null) {
+        // first occurrence in base
+        hashOccurrence = new HashOccurrence();
+        hashOccurrence.baseLine = line;
+        hashOccurrence.baseCount = 1;
+        occurrencesByHash.put(hash, hashOccurrence);
+      } else {
+        hashOccurrence.baseCount++;
+      }
+    }
+
+    for (Integer line : rawsByLine.keySet()) {
+      int hash = rawHashSequence.getBlockHashForLine(line);
+      HashOccurrence hashOccurrence = occurrencesByHash.get(hash);
+      if (hashOccurrence != null) {
+        hashOccurrence.rawLine = line;
+        hashOccurrence.rawCount++;
+      }
+    }
+
+    for (HashOccurrence hashOccurrence : occurrencesByHash.values()) {
+      if (hashOccurrence.baseCount == 1 && hashOccurrence.rawCount == 1) {
+        // Guaranteed that baseLine has been moved to rawLine, so we can map all issues on baseLine to all issues on rawLine
+        map(rawsByLine.get(hashOccurrence.rawLine), basesByLine.get(hashOccurrence.baseLine), tracking);
+        basesByLine.removeAll(hashOccurrence.baseLine);
+        rawsByLine.removeAll(hashOccurrence.rawLine);
+      }
+    }
+
+    // Check if remaining number of lines exceeds threshold
+    if (basesByLine.keySet().size() * rawsByLine.keySet().size() < 250000) {
```

Comment:
this condition would better be a method and the number a constant, would save the comment